samples/directx/d3d10_opencl.cpp

```diff
@@ -0,0 +1,397 @@
+        scd.OutputWindow      = m_hWnd;                          // the window to be used
+        scd.SampleDesc.Count  = 1;                               // how many multisamples
+        scd.Windowed          = TRUE;                            // windowed/full-screen mode
+        scd.SwapEffect        = DXGI_SWAP_EFFECT_DISCARD;
+        scd.Flags             = DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH; // allow full-screen switching
+
+        r = ::D3D10CreateDeviceAndSwapChain(
+                NULL,
+                D3D10_DRIVER_TYPE_HARDWARE,
+                NULL,
+                0,
+                D3D10_SDK_VERSION,
+                &scd,
+                &m_pD3D10SwapChain,
+                &m_pD3D10Dev);
+        if (FAILED(r))
+        {
+            return -1;
+        }
+
+        r = m_pD3D10SwapChain->GetBuffer(0, __uuidof(ID3D10Texture2D), (LPVOID*)&m_pBackBuffer);
+        if (FAILED(r))
+        {
+            return -1;
+        }
+
+        r = m_pD3D10Dev->CreateRenderTargetView(m_pBackBuffer, NULL, &m_pRenderTarget);
+        if (FAILED(r))
+        {
+            return -1;
+        }
+
+        m_pD3D10Dev->OMSetRenderTargets(1, &m_pRenderTarget, NULL);
+
+        D3D10_VIEWPORT viewport;
+        ZeroMemory(&viewport, sizeof(D3D10_VIEWPORT));
+
+        viewport.Width    = m_width;
+        viewport.Height   = m_height;
+        viewport.MinDepth = 0.0f;
+        viewport.MaxDepth = 0.0f;
+
+        m_pD3D10Dev->RSSetViewports(1, &viewport);
+
+        D3D10_TEXTURE2D_DESC desc = { 0 };
+
+        desc.Width            = m_width;
+        desc.Height           = m_height;
+        desc.MipLevels        = 1;
+        desc.ArraySize        = 1;
+        desc.Format           = DXGI_FORMAT_R8G8B8A8_UNORM;
+        desc.SampleDesc.Count = 1;
+        desc.BindFlags        = D3D10_BIND_SHADER_RESOURCE;
+        desc.Usage            = D3D10_USAGE_DYNAMIC;
+        desc.CPUAccessFlags   = D3D10_CPU_ACCESS_WRITE;
+
+        r = m_pD3D10Dev->CreateTexture2D(&desc, NULL, &m_pSurface);
+        if (FAILED(r))
+        {
+            std::cerr << "Can't create texture with input image" << std::endl;
+            return -1;
+        }
+
+        if (cv::ocl::haveOpenCL())
+        {
+            m_oclCtx = cv::directx::ocl::initializeContextFromD3D10Device(m_pD3D10Dev);
+        }
+
+        m_oclDevName = cv::ocl::useOpenCL() ?
+            cv::ocl::Context::getDefault().device(0).name() :
+            "No OpenCL device";
+
+        return 0;
+    } // init()
+
+
+    int get_surface(ID3D10Texture2D** ppSurface)
+    {
+        HRESULT r;
+
+        if (!m_cap.read(m_frame_bgr))
+            return -1;
+
+        cv::cvtColor(m_frame_bgr, m_frame_rgba, CV_RGB2BGRA);
+
+        UINT subResource = ::D3D10CalcSubresource(0, 0, 1);
+
+        D3D10_MAPPED_TEXTURE2D mappedTex;
+        r = m_pSurface->Map(subResource, D3D10_MAP_WRITE_DISCARD, 0, &mappedTex);
```

Comment:
There are multiple surface Map/Unmap calls during single `render()` call.
May be build pipeline that works with UMat instead of texture surface?